gateway/cert_test.go

```diff
@@ -1732,4 +1807,257 @@ func TestStaticMTLSAPI(t *testing.T) {
 			Code:      http.StatusOK,
 		})
 	})
+
+	t.Run("validate client cert against certificate authority", func(t *testing.T) {
+		rootCertPEM, rootKeyPEM, err := generateRootCertAndKey(t)
+		assert.NoError(t, err)
+
+		serverCertPEM, serverKeyPEM, err := generateServerCertAndKeyChain(t, rootCertPEM, rootKeyPEM)
+		assert.NoError(t, err)
+		combinedPEM := bytes.Join([][]byte{serverCertPEM.Bytes(), serverKeyPEM.Bytes()}, []byte("\n"))
+
+		certID, _, err := certs.GetCertIDAndChainPEM(combinedPEM, "")
+		assert.NoError(t, err)
+
+		conf := func(globalConf *config.Config) {
+			globalConf.Security.ControlAPIUseMutualTLS = false
+			globalConf.HttpServerOptions.UseSSL = true
+			globalConf.HttpServerOptions.SSLInsecureSkipVerify = false
+			globalConf.HttpServerOptions.SSLCertificates = []string{"default" + certID}
+			globalConf.SuppressRedisSignalReload = true
+		}
+		ts := StartTest(conf)
+
+		certID, err = ts.Gw.CertificateManager.Add(combinedPEM, "default")
+		assert.NoError(t, err)
+		defer ts.Gw.CertificateManager.Delete(certID, "default")
+		ts.ReloadGatewayProxy()
+
+		clientCertPEM, clientKeyPEM, err := generateClientCertAndKeyChain(t, rootCertPEM, rootKeyPEM)
+		assert.NoError(t, err)
+
+		rootCertID, err := ts.Gw.CertificateManager.Add(rootCertPEM, "default")
+		assert.NoError(t, err)
+
+		clientCert, _ := tls.X509KeyPair(clientCertPEM.Bytes(), clientKeyPEM.Bytes())
+
+		ts.Gw.BuildAndLoadAPI(func(spec *APISpec) {
+			spec.APIID = "apiID-1"
+			spec.UseMutualTLSAuth = true
+			spec.Proxy.ListenPath = "/static-mtls"
+			spec.ClientCertificates = []string{rootCertID}
+		})
+
+		t.Run("valid client", func(t *testing.T) {
+			validCertClient := GetTLSClient(&clientCert, rootCertPEM)
+			_, _ = ts.Run(t, test.TestCase{
+				Domain: "localhost",
+				Client: validCertClient,
+				Path:   "/static-mtls",
+				Code:   http.StatusOK,
+			})
+		})
+
+		t.Run("invalid client", func(t *testing.T) {
+			_, _, _, invalidClientCert := crypto.GenCertificate(&x509.Certificate{}, false)
+			tlsConfig := GetTLSConfig(&invalidClientCert, nil)
+			tlsConfig.InsecureSkipVerify = false
+			transport := &http.Transport{TLSClientConfig: tlsConfig}
+
+			invalidClient := &http.Client{Transport: transport}
+			u, err := url.Parse(ts.URL)
+
+			req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("localhost:%s/static-mtls", u.Port()), nil)
+			assert.NoError(t, err)
+			_, err = invalidClient.Do(req)
+			assert.Error(t, err)
```

Comment:
We may need to have a stronger assertion for the error returned here, so we know it's related to TLS. I asume the request could fail due to some routing related issue (invalid host, rate limit, etc.).

Reply:
@jeffy-mathew unresolved